ruskeyverbs/models.py

```diff
@@ -26,6 +29,42 @@ class Verb(models.Model):
     def __str__(self):
         return self.infinitive
 
+    def get_earliest_due_date(self, my_user):
```

Comment:
This whole method is really pretty terrible from a sql performance standpoint. But the first step is get it working and the next step is optimization. So I'll try and go through and share my thoughts.

The point of this method is to get the earliest due date for a given user and verb across all examples for that verb. I hope I'm understanding that correctly. If I am then I see two major, closely related issues with how this method is written. The first is the number of sql queries running this method will cause to be run in the db.

It's important to understand that creating a queryset object in memory does NOT hit the database at all. Querysets are simply logical representations of a query. In line 33 you grab a reference to one (self.example_set.all()) and set it to the variable 'examples'. This does not cause a query to be executed. At this point, examples is simply a logical representation of the notion of all the examples related to the current verb.

The query that 'examples' represents is not generated nor executed until line 36, when you iterate over the results. At that point, a query very similar to 'SELECT * FROM Examples WHERE verb_id = {x}' is compiled by the Django ORM and then sent to the db to be executed. The results are then used to populate a list of Example instances.

So far there isn't a problem here. The problem arises in line 38 where you now need to get the PerformancePerExample for that example for the given user. You not only create the queryset on line 38, but you also iterate it with the array indexer, accessing the first record returned, which causes the query to be compiled and run in the database, the results returned, a PerformancePerExample instance being created and populated, and finally the due_date being accessed and stored in the 'my_due_date' variable.

The issue is that this query will need to be run N times, where N is the number of examples for the current verb. Running a query in the db is an expensive operation. A connection to the database needs to be established, the queryset needs to be compiled down to SQL, the SQL needs to be sent to the db, the DB needs to parse the SQL and return a result set, and finally Django needs to take that result set and turn it into objects and then close the db connection.

That's a lot of work for every single query you run, so you really want to keep your queries to a minimum where possible.

And that leads me to the second, related problem with this method: The operation you're doing here is really what databases are made for. Yet you're doing most of the work in python, in memory. Databases are great at tying related data together and then sorting and aggregating that data. And that's what you're doing here. You're traversing a data relationship and then doing an aggregate function, the Min(imum) function, on the due_date field in the related dataset.

I'm pretty sure that the exact same result can be arrived at with a single query and a single line of code, like so:

min_due_date = PerformancePerExample.objects.filter(user=my_user, example__verb=self).aggregate(Min(
'due_date'))

It's much more efficient when it runs and it's way easier to read and maintain. Think about it and let me know what questions you have.

Reply:
Thank you for the great explanation. That is very helpful. I had spent some time trying to figure out how to do this in a single query, but I got kind of bogged down even after our email exchange on the subject. Having the costs of my approach laid out here is really helpful. I am going to read through the query and aggregation documentation in its entirety before I fix this.